### PR TITLE
mysql proxy: connection attributes parsing 

### PR DESCRIPTION
**Pull request by qinggniq**
Signed-off-by: qinggniq <livewithblank@gmail.com>

Parse connection attributes when CLIENT_CONNECT_ATTRS is set at the client capability.

[Some connector](https://dev.mysql.com/doc/refman/8.0/en/performance-schema-connection-attribute-tables.html) will set connection attributes when connecting. If CLIENT_CONNECT_ATTRS is set but the related field is empty, the server will return bad handshake error. 

Commit Message: MySQL proxy codec parsing of connection attributes 
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by qinggniq** [comment]
@cpakulski PTAL.

File: source/extensions/filters/network/mysql_proxy/mysql_codec_clogin.cc [hunk 1]
```diff
@@ -155,6 +162,49 @@ DecodeStatus ClientLogin::parseResponse41(Buffer::Instance& buffer) {
     ENVOY_LOG(debug, "error when parsing auth plugin name of client login message");
     return DecodeStatus::Failure;
   }
+  if (client_cap_ & CLIENT_CONNECT_ATTRS) {
+    // length of all key value pairs
+    uint64_t kvs_len;
+    if (BufferHelper::readLengthEncodedInteger(buffer, kvs_len) != DecodeStatus::Success) {
+      ENVOY_LOG(debug, "error when parsing length of all key-values in connection attributes of "
```
Comment (cpakulski) [inline source/extensions/filters/network/mysql_proxy/mysql_codec_clogin.cc:169]: Rephrase to 
when parsing _total_ length of key values ...

File: source/extensions/filters/network/mysql_proxy/mysql_codec_clogin.cc [hunk 2]
```diff
@@ -155,6 +162,49 @@ DecodeStatus ClientLogin::parseResponse41(Buffer::Instance& buffer) {
...
+      if (BufferHelper::readStringBySize(buffer, str_len, val) != DecodeStatus::Success) {
+        ENVOY_LOG(debug, "error when parsing connection attribute val in connection attributes of "
+                         "client login message");
+        return DecodeStatus::Failure;
+      }
+      conn_attr_.emplace_back(std::make_pair(std::move(key), std::move(val)));
+      kvs_len -= prev_len - buffer.length();
+    }
+  }
+  ENVOY_LOG(debug, "parse client login protocol 41, consumed len {}, remain {}",
```
Comment (cpakulski) [inline source/extensions/filters/network/mysql_proxy/mysql_codec_clogin.cc:206]: change to 
parsed client

File: test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc [hunk 3]
```diff
@@ -553,6 +556,180 @@ TEST_F(MySQLCLoginTest, MySQLClientLogin41IncompleteAuthPluginName) {
   EXPECT_EQ(mysql_clogin_decode.getAuthPluginName(), "");
 }
 
+/*
+ * Negative Test the MYSQL Client Login 41 message parser:
+ * Incomplete length of connection attributions
+ */
+TEST_F(MySQLCLoginTest, MySQLClientLogin41IncompleteConnAttrLength) {
```
Comment (cpakulski) [inline test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc:563]: There is lost of repeated code in next 5 tests. Consider maybe deriving another class from _MySQLCLoginTest_, for example _MySQL41LoginTest_ and move the repeated code to its methods.

**Review by cpakulski — CHANGES_REQUESTED** [review]
Overall looks good. 
Some spelling mistakes and tests should avoid so much duplicated code.

File: source/extensions/filters/network/mysql_proxy/mysql_codec_clogin.cc (hunk 1 above)
Comment (qinggniq) [reply source/extensions/filters/network/mysql_proxy/mysql_codec_clogin.cc:169]: fixed.

File: source/extensions/filters/network/mysql_proxy/mysql_codec_clogin.cc (hunk 2 above)
Comment (qinggniq) [reply source/extensions/filters/network/mysql_proxy/mysql_codec_clogin.cc]: fixed.

File: test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc (hunk 3 above)
Comment (qinggniq) [reply test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc]: fixed.

Comment (cpakulski) [reply test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc]: Looks much better!

File: test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc [hunk 4]
```diff
@@ -553,6 +558,106 @@ TEST_F(MySQLCLoginTest, MySQLClientLogin41IncompleteAuthPluginName) {
...
+    EXPECT_EQ(login_decode.getDb(), login_encode.getDb());
+    EXPECT_EQ(login_decode.getAuthPluginName(), login_encode.getAuthPluginName());
+    if (additional_check) {
+      additional_check();
+    }
+  }
+
+  ClientLogin login_encode;
+  ClientLogin login_decode;
+  int incomplete_base_len;
```
Comment (cpakulski) [inline test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc:598]: add trailing _ to data members.

File: test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc [hunk 5]
```diff
@@ -553,6 +558,106 @@ TEST_F(MySQLCLoginTest, MySQLClientLogin41IncompleteAuthPluginName) {
...
+    EXPECT_TRUE(login_decode.isConnectWithDb());
+    EXPECT_EQ(login_decode.getClientCap(), login_encode.getClientCap());
+    EXPECT_EQ(login_decode.getExtendedClientCap(), login_decode.getExtendedClientCap());
+    EXPECT_EQ(login_decode.getMaxPacket(), login_encode.getMaxPacket());
+    EXPECT_EQ(login_decode.getCharset(), login_encode.getCharset());
+    EXPECT_EQ(login_decode.getUsername(), login_encode.getUsername());
+    EXPECT_EQ(login_decode.getAuthResp(), login_encode.getAuthResp());
+    EXPECT_EQ(login_decode.getDb(), login_encode.getDb());
+    EXPECT_EQ(login_decode.getAuthPluginName(), login_encode.getAuthPluginName());
+    if (additional_check) {
```
Comment (cpakulski) [inline test/extensions/filters/network/mysql_proxy/mysql_clogin_test.cc:591]: Can you compare to nullptr?

File: test/extensions/filters/network/mysql_proxy/mysql_test_utils.cc [hunk 6]
```diff
@@ -122,6 +122,18 @@ std::string MySQLTestUtils::encodeMessage(uint32_t packet_len, uint8_t it, uint8
   return buffer.toString();
 }
 
+int MySQLTestUtils::bytesOfConnAtrributeLength(
+    const std::vector<std::pair<std::string, std::string>> conn_attrs) {
+  int64_t allLen = 0;
```
Comment (cpakulski) [inline test/extensions/filters/network/mysql_proxy/mysql_test_utils.cc:127]: rename to all_len.

**Review by cpakulski — CHANGES_REQUESTED** [review]
The last small comments.

**Review by cpakulski — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**